### PR TITLE
Simplify cli-handling functions

### PR DESCRIPTION
**Pull request by matklad**
This simplifies somewhat roundabout `main` in `bin/cargo.rs`.

### CODE REVIEW

File: src/cargo/lib.rs [hunk 1]
```diff
@@ -72,27 +59,7 @@ pub fn call_main_without_stdin<T, V>(
...
-        callback(&args, config.as_ref().unwrap())
-    })();
-    let mut verbose_shell = shell(Verbose, Auto);
-    let mut shell = config.as_ref().map(|s| s.shell());
-    let shell = shell.as_mut().map(|s| &mut **s).unwrap_or(&mut verbose_shell);
-    process_executed(result, shell)
-}
-
-pub fn process_executed<T>(result: CliResult<Option<T>>, shell: &mut MultiShell)
+pub fn process_executed<T>(result: CliResult<Option<T>>, shell: &mut MultiShell) -> !
```
Comment (matklad) [inline src/cargo/lib.rs]: This is interesting: previously, this function diverged for `Err`s and returned for `Ok`s. I've made it an always diverging function, and this highlights that sometimes we may leave some stuff on the stack undropped.

File: src/bin/cargo.rs [hunk 2]
```diff
@@ -69,7 +68,31 @@ See 'cargo help <command>' for more information on a specific command.
...
+            })
+        }).collect());
+        let rest = &args;
+        cargo::call_main_without_stdin(execute, &config, USAGE, rest, true)
+    })();
+
+    match result {
+        Err(e) => cargo::handle_cli_error(e, &mut *config.shell()),
+        Ok(None) => {},
+        Ok(Some(())) => unreachable!(),
```
Comment (matklad) [inline src/bin/cargo.rs]: This case is here because we have a great infrastrucure for commands to return `Encodable` instances to be printing in the JSON format. 

In reality, only three commands use this feature (deprecated `read_manifest`, `verify_project` which prints `Map<String, String>` of errors and `metadata`). Everyone else is forced to return the funny `Option<()>` type. I want to strip this infrastructure in the next PR and instead make the three commands to print JSON directly.

**Comment by alexcrichton** [comment]
Looks good to me, thanks! These all date back to basically the first lines of code written in Cargo, so they can all definitely use an update :)

r=me with a rebase

**Comment by matklad** [comment]
Hm, something iffy is going on (in the master branch of Cargo, and in this PR as well)! [Here](https://github.com/rust-lang/cargo/blob/8d5383866e387a5bf88e25181d23c1af67cb3abe/src/bin/cargo.rs#L128) we create a job object on windows, which does some useful stuff in Drop. And [on this line](https://github.com/rust-lang/cargo/blob/8d5383866e387a5bf88e25181d23c1af67cb3abe/src/bin/cargo.rs#L201) we leak it, if the command executes with an error, because `cargo::process_executed` diverges in this case. 

So together with rebasing I have restored the original behavior of `process_executed`. I'll look into this separately.

File: tests/lockfile-compat.rs [hunk 3]
```diff
@@ -1,4 +1,3 @@
-#[macro_use]
```
Comment (matklad) [inline tests/lockfile-compat.rs]: Just a drive-by warning fix.

**Comment by alexcrichton** [comment]
Hm fascinating! The destructor for that job object was definitely intended to always run, both in the successful and failing case. That's what helps us gracefully tear down processes on Windows rather than abruptly.

@bors: r+

**Comment by sfackler** [comment]
This change adds a decent bit of boilerplate to third party subcommands: https://github.com/sfackler/cargo-sls-distribution/commit/05f0937e92e17c06378b68da3ba4c2d6979475c0#diff-639fbc4ef05b315af92b4d836c31b023R81

**Comment by matklad** [comment]
Interesting! It didn't occur to me that external subcommands link to cargo library directly...

Why do you need to link with cargo directly though? It looks like you are executing `cargo build --bin` and then do something with the resulting artifacts. Perhaps it would be simpler to call cargo as a subprocess from the subcommand, and them process the result?

**Comment by sfackler** [comment]
For that specific subcommand, I'm not aware of a way to track down the resulting artifacts when shelling out to cargo. More generally though, other subcommands like https://github.com/sfackler/cargo-tree need to integrate more tightly into cargo's internals.

There are quite a lot of subcommands that link into cargo as a library: https://crates.io/crates/cargo/reverse_dependencies

**Comment by matklad** [comment]
Hm, looks like `cargo-tree` visualizes the graph one can get from `cargo metadata`, no? 

@alexcrichton I wonder what do you think about stability of cargo as a library, especially given that you had to update `cargo vendor` recently?

On the one hand, we should not have broken a ton of external subcommands, on the other hand, I still don't like command related API and still want to replace it with a less copy-pasty clap one day :) 

What we certainly can do is write some more docs about how json information can be used to write custom subcommands...

**Comment by alexcrichton** [comment]
Oh I consider the `cargo` crate itself wildly unstable, we'll basically never issue a point release (only ever major version releases) for the near future (aka 5 year timescale). That being said I think it's also quite handy to use the `cargo` crate itself with programmatic access as it gives you tons of internals the command line doesn't. In that sense we shouldn't bend over backwards to accomodate external crates but we also shouldn't bend over backwards to break external crates.

I think `cargo metadata` has made a lot of progress to *nowadays* implement `cargo tree` with (when `cargo tree` as added I don't think we had a `metadata` command). I see `cargo metadata` as the go-to "I'd like to write a tool on cargo" format. It's guaranteed stable and you don't have to bring in all of the cargo crate (which is quite large). If you need more information than `cargo metadata` gives you then I'd consider it ok to link to Cargo itself.

In the end I could go either way on this, I don't mind copy/pasting a few lines from `src/bin/cargo.rs` into external commands :)
